KOGITO-1301: Adding support for binary icons

jesuino opened the pull request:
KOGITO-1301: Adding support for icons in Work Items
--

In this change we are retrieving binary content, creating the URL for a base64 content and returning a default ICON if the icon could not be loaded. The result can be seen below:

![work_items_definitions_with_icons](https://user-images.githubusercontent.com/359820/75467431-64287580-596a-11ea-98f4-0dbcafcc653f.gif)


PART OF AN ENSEMBLE, Merge with:
https://github.com/kiegroup/kogito-tooling/pull/68

jesuino commented:
Jenkins execute full downstream build

jesuino commented:
It also works on chrome now, the missing icons are replaced by the default icon

![image](https://user-images.githubusercontent.com/359820/75480686-744b4f80-5980-11ea-8c4b-d13a978e91f3.png)

jesuino commented:
Jenkins execute full downstream build

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-kogito-runtime/src/main/java/org/kie/workbench/common/stunner/kogito/client/services/WorkItemDefinitionStandaloneClientService.java:
```diff
@@ -46,6 +48,21 @@
 @ApplicationScoped
 public class WorkItemDefinitionStandaloneClientService implements WorkItemDefinitionClientService {
 
+    private static final String DEFAULT_ICON_DATA = "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABEAAAARCAYAAAA7bUf6AAADVklEQVQ4EYVUbWhTVxg+UAR/" +
```
Some comments about the "default" icon data:

The initial idea was that the `WorkItemDefinition` (domain) object has a field `iconData`, but I can be empty. This way:
* Once parsing a .wid file, if there exist icon, the `iconData` will get populated with the properly value
* Otherwise, if an icon cannot be loaded (for whatever reason), the `iconData` field results empy. This way:
** All the components that display those icons (palette, canvas elemnts, etc) are actually taking into account the empty  data case, and they should show some default icon data
** It also implies that the javascript object does not have the same default icon data value accross instances, we just keep the objec'ts field empty, and then components can react to it 

It means there should be no need to set any default icon data, in case the icon is not properly loaded, just keep the String empty or `null`, and the different Stunner components woulld prpoerly show some default icon. Didn't that worked? 

Also, I would say there is no need for this field, even in case we need it for setting some default value, because the default icon data is already being provided by a call to this static method - [WorkItemDefinitionClientUtils#getDefaultIconData()](https://github.com/kiegroup/kie-wb-common/blob/1d96fdcb991db06a633f620e5b53a4dd22c0b5c9/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/workitem/WorkItemDefinitionClientUtils.java#L25)

jesuino replied:
Hello Roger!

Thank you so much for the feedback.

I removed the default icon and with the changes everything seen to be working now, could you please check:

https://github.com/kiegroup/kie-wb-common/pull/3195/commits/5f61a4fb051ddd2ceefe352e543af453e91bfed3

What I had to do:

* Build the image data URI (https://en.wikipedia.org/wiki/Data_URI_scheme)
* Ask resource content API for binary content 
* The fixes on kogito-tooling side.

Thanks

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-kogito-runtime/src/main/java/org/kie/workbench/common/stunner/kogito/client/services/WorkItemDefinitionStandaloneClientService.java:
```diff
@@ -131,43 +148,57 @@ public void destroy() {
+                                             if (nonEmpty(iconData)) {
+                                                 wid.getIconDefinition().setIconData(iconDataUri(iconUri, iconData));
+                                             }
+                                             return promises.resolve(DEFAULT_ICON_DATA);
```
I think this is not correct, although maybe it has no implications on the final result... I mean, if exist some `iconData`, we'll set the value to the model object, but we also sholud return that content as a result for the promise, isn't it?

I mean, actually, you're setting the right `iconData` to the object model, and that will make things work, but if someone using this API uses the promise result, it would always obtain the default `iconData`, instead of the actual one, isn't it?

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-kogito-runtime/src/main/java/org/kie/workbench/common/stunner/kogito/client/services/WorkItemDefinitionStandaloneClientService.java:
```diff
@@ -131,43 +148,57 @@ public void destroy() {
+            return "data:image/" + fileType + ";base64, " + iconData;
         }
-        return promises.resolve();
+        return DEFAULT_ICON_DATA;
```
same comment as above, althogugh it has no impact on the resutls...

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-kogito-runtime/src/main/java/org/kie/workbench/common/stunner/kogito/client/services/WorkItemDefinitionStandaloneClientService.java:
```diff
@@ -131,43 +148,57 @@ public void destroy() {
+                                                         })
+                                                         .catch_(error -> {
+                                                             failure.onInvoke(error);
+                                                             return null;
```
As far as I tested this, by retuning `null` for a promise call, it end up on NPE's in the unit tests... did you tested unit tests still work properly after it?

domhanak replied:
@jesuino is this Question resolved?

jesuino replied:
Hello,

Yes, it is fine, no NPE on tests when I checked!

romartin commented:
Should somebody from QE also verify this? CC @LuboTerifaj @domhanak ? Thx in advance!

jesuino commented:
@romartin 

Thanks for the feedack. I ran the tests, no NPE, but I added tests for the methods I created and removed a leftover manual debug line :)

jesuino commented:
Jenkins execute full downstream build

jesuino commented:
jenkins retest this

jesuino commented:
Jenkins execute full downstream build

jesuino commented:
jenkins retest this

jesuino commented:
jenkins retest this

jesuino commented:
Jenkins execute full downstream build

romartin commented:
Jenkins retest this

jesuino commented:
Hello @romartin I pushed more changes earlier today, did your build included those? Today's morning I notice some issues with the VSCode part and worked with @tiagobento to fix it, see:
https://github.com/kiegroup/kogito-tooling/pull/68/commits/120281175b1a48710b4a6830b73535f02f1fd830

If yes, I will re-test tomorrow to make sure everything is working.

romartin commented:
Hey @jesuino 

Just rebased from master now, cherry-picked your commits on top, and built the plugin... still cannot open the process, in case some icons are missing or when loading some icon...

For example, I've a .wid file that declares several service tasks, but only added the icon for one in the project root folder, so it should be displaying the right icon for that service task, and the default icon for others... but the editor still does'nt open:

![Screenshot from 2020-03-10 02-03-30](https://user-images.githubusercontent.com/4602417/76269897-afe4f400-6273-11ea-9640-5845fef34b12.png)

PS: This time I'm using the right snapshot version and even introduced some loggers to ensure the plugin was running my local updates, so I think something is still missing to make this work properly, all steps from my side seem to be correct, please lemme know if I'm missing something.

romartin commented:
Hey @domhanak ! this PR is starting to be urgent, could u plz try to take it a look? Thx in advance!

domhanak commented:
LGTM, feel free to merge when green builds

domhanak commented:
jenkins retest this

romartin commented:
great, merging, thanks guys! :+1:
